game-core/src/main/java/games/strategy/triplea/ai/pro/AbstractProAi.java

```diff
@@ -86,24 +90,35 @@ public AbstractProAi(
     storedPurchaseTerritories = null;
     storedPoliticalActions = null;
     storedStrafingTerritories = new ArrayList<>();
+
+    // in case there are multiple Pro AIs running at the same time, unregister the one
+    // created before creating one so that there is only one menu item for all of the ProAI
+    DebugMenu.unregisterMenuCallback("Hard AI");
```

Comment:
Could `registerMenuCallback` handle this detail for us?

Reply:
Second, it's problematic that AI now has a dependency on swing. AI should ideally not depend on the UI layer at all.

For example, we saw this as a major problem when trying to replace Swing with JavaFX.

Presumably this could be fixed by passing a `DebugMenu` interface to AbstractProAi which it could use to send data but not be aware of the UI implementation detail. WDYT @trevan ?

Reply:
So which code would own the `ProLogAi`?  That looks like it should be AI code since it is only used by AI players.  And the `ProLogAi` can't be made unaware of the UI since it has buttons, checkboxes, etc.  It is a dialog that it needs to render.

Trying to split the debug menu items and their actions from swing or javafx feels like it will just require us to create a new ui library.

Reply:
I would actually expect `registerMenuCallback` to throw an exception if a duplicate string is registered.  The duplicate "Hard AI" is a relic of how Hard AI groups everything together in one logging ui.